Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact
the validity of the results.

 Assumptions of ANOVA and Examples of Violations

Assumptions of ANOVA are:

Normality: The dependent variable is normally distributed within each group.
Homogeneity of Variance: The variance of the dependent variable is equal across all groups.
Independence: Observations within each group are independent of each other.
Random Sampling: Samples are randomly selected from the population.
Examples of violations that could impact the validity of the results are:

Non-normality: If the distribution of the dependent variable is significantly non-normal, the results of ANOVA may not be valid. For example, if the distribution is skewed or has outliers.
Heterogeneity of Variance: If the variance of the dependent variable is not equal across all groups, the results of ANOVA may not be valid. For example, if one group has a much larger variance than the other groups.
Dependence: If observations within each group are not independent of each other, the results of ANOVA may not be valid. For example, if the same subjects are used in multiple groups.
Non-Random Sampling: If samples are not randomly selected from the population, the results of ANOVA may not be valid. For example, if samples are selected based on convenience or preference.

Q2. What are the three types of ANOVA, and in what situations would each be used?

The three types of ANOVA are:

One-Way ANOVA: Used to compare means of a dependent variable across two or more independent groups.
Two-Way ANOVA: Used to compare means of a dependent variable across two or more independent groups while controlling for the effects of two independent variables.
Repeated Measures ANOVA: Used to compare means of a dependent variable across two or more related groups over time.

Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

The partitioning of variance in ANOVA is the process of decomposing the total variance of the dependent variable into different sources. The total variance is divided into two parts: the variance due to differences between the groups (explained variance) and the variance due to differences within the groups (unexplained variance or error variance). 

Understanding the partitioning of variance is important because it allows us to determine the proportion of variance that is explained by the independent variable and to test the significance of the independent variable.

Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual
sum of squares (SSR) in a one-way ANOVA using Python?

In [2]:
import pandas as pd
from statsmodels.formula.api import ols
import seaborn as sns
from statsmodels.stats.anova import anova_lm

# Loading Iris dataset from seaborn
df_iris = sns.load_dataset('iris')
print('Top 5 rows of IRIS dataset : ')
print(df_iris.head())
print('\n===================================================================\n')

# Fit the one-way ANOVA model (sepal length vs Species)
model = ols('sepal_length ~ species', data=df_iris).fit()

# Calculate the sum of squares for the model
print('Values for Sepal Length vs Species:')
SSE = model.ess
SSR = model.ssr
SST = SSE + SSR

print('SSE:', round(SSE,4))
print('SSR:', round(SSR,4))
print('SST:', round(SST,4))

print('\n===================================================================\n')
# Print the ANOVA table
print(anova_lm(model))

Top 5 rows of IRIS dataset : 
   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa


Values for Sepal Length vs Species:
SSE: 63.2121
SSR: 38.9562
SST: 102.1683


             df     sum_sq    mean_sq           F        PR(>F)
species     2.0  63.212133  31.606067  119.264502  1.669669e-31
Residual  147.0  38.956200   0.265008         NaN           NaN


Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

Top 5 rows of Tooth Growth Dataset
    len supp  dose
0   4.2   VC   0.5
1  11.5   VC   0.5
2   7.3   VC   0.5
3   5.8   VC   0.5
4   6.4   VC   0.5


Main effects:
C(supp)     205.350000
C(dose)    2426.434333
Name: sum_sq, dtype: float64


Interaction effect:
C(supp):C(dose)    108.319
Name: sum_sq, dtype: float64


ANOVA Table:
                      sum_sq    df          F        PR(>F)
C(supp)           205.350000   1.0  15.571979  2.311828e-04
C(dose)          2426.434333   2.0  91.999965  4.046291e-18
C(supp):C(dose)   108.319000   2.0   4.106991  2.186027e-02
Residual          712.106000  54.0        NaN           NaN


Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02.
What can you conclude about the differences between the groups, and how would you interpret these
results?

With an F-statistic of 5.23 and a p-value of 0.02, we can conclude that there are significant differences between the groups being compared. The F-statistic is a ratio of between-group variance to within-group variance, with higher values indicating greater between-group differences relative to within-group differences. The p-value indicates the probability of obtaining an F-statistic as extreme or more extreme than the one observed, assuming the null hypothesis (i.e., no group differences) is true. In this case, with a p-value of 0.02, we reject the null hypothesis and conclude that there are significant group differences. The effect size should also be considered to determine the practical significance of the findings.

Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential
consequences of using different methods to handle missing data?

In a repeated measures ANOVA, missing data can be handled using various methods such as pairwise deletion, listwise deletion, imputation, and maximum likelihood estimation. Pairwise deletion retains data for each individual variable and omits data from any cases with missing data. Listwise deletion omits entire cases that contain missing data, which can result in a loss of statistical power and introduce bias if the data are not missing at random. Imputation involves estimating the missing values using statistical techniques such as mean imputation, regression imputation, or multiple imputation. Maximum likelihood estimation involves estimating the model parameters while accounting for the missing data using the observed data and an assumption about the distribution of the missing data.

The potential consequences of using different methods to handle missing data include biased estimates of the model parameters, reduced statistical power, and increased risk of type I and type II errors. The choice of method depends on the pattern and amount of missing data, the research question, and assumptions about the missing data mechanism.

Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide
an example of a situation where a post-hoc test might be necessary.

Common post-hoc tests used after ANOVA include the Tukey-Kramer test, the Scheffe test, the Bonferroni test, and the Dunnett test. The Tukey-Kramer test is used to determine which specific groups differ significantly from each other in pairwise comparisons. The Scheffe test is a more conservative test that controls for familywise error rate (FWER) and can be used when the number of pairwise comparisons is large. The Bonferroni test is another conservative test that controls for the FWER by adjusting the significance level for each comparison. The Dunnett test is used to compare multiple groups to a control group, with less stringent assumptions than the other post-hoc tests.

An example of a situation where a post-hoc test might be necessary is when conducting a study comparing the effectiveness of three different treatments for a medical condition. After conducting an ANOVA and finding a significant overall effect, a post-hoc test can be used to determine which specific treatments are significantly different from each other.

Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from
50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python
to determine if there are any significant differences between the mean weight loss of the three diets.
Report the F-statistic and p-value, and interpret the results.

In [9]:
import numpy as np
from scipy.stats import f_oneway

# Generate simulated data assuming normal distribution with same variance
np.random.seed(1)
diet_A = np.random.normal(5, 1, 50)
diet_B = np.random.normal(4, 1, 50)
diet_C = np.random.normal(3, 1, 50)

# Perform one-way ANOVA
f_statistic, p_value = f_oneway(diet_A, diet_B, diet_C)

# Set significance level
alpha = 0.05

# Null hypothesis: The mean weight loss is the same for all three diets.
# Alternative hypothesis: The mean weight loss is different for at least one diet.
null_hypothesis = "The mean weight loss is the same for all three diets."
alternate_hypothesis = "The mean weight loss is different for at least one diet."

print("F-statistic:", f_statistic)
print("p-value:", p_value)
if p_value < alpha:
    print("We reject the null hypothesis.")
    print(f"Conclusion : {alternate_hypothesis}")
else:
    print("We fail to reject the null hypothesis.")
    print(f"Conclusion : {null_hypothesis}")

F-statistic: 57.06379442059458
p-value: 4.5619061215783055e-19
We reject the null hypothesis.
Conclusion : The mean weight loss is different for at least one diet.


In [10]:
#Performing Tukey's test for mean difference
import numpy as np
from statsmodels.stats.multicomp import pairwise_tukeyhsd

#create DataFrame to hold data
df = pd.DataFrame({'weight_loss': list(diet_A) + list(diet_B) + list(diet_C),
                   'group': np.repeat(['A', 'B', 'C'], repeats=50)})

# perform Tukey's test
tukey = pairwise_tukeyhsd(endog=df['weight_loss'],
                          groups=df['group'],
                          alpha=0.05)

# Print results
print(tukey)

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower   upper  reject
---------------------------------------------------
     A      B  -0.8278   0.0 -1.2477 -0.4079   True
     A      C  -1.8898   0.0 -2.3097 -1.4699   True
     B      C   -1.062   0.0 -1.4819 -0.6421   True
---------------------------------------------------


Above interpretation means all three means are different reject value is True for all of 3
Mean Difference between diet_A and diet_B is -0.8278
Mean Difference between diet_A and diet_C is -1.8898
Mean Difference between diet_A and diet_C is -1.062
Maximum mean difference is in between diet_A and diet_C
All mean differences with diet_A are negative
diet_A has shown highest weight loss compared to diet_B and diet_C


Q10. A company wants to know if there are any significant differences in the average time it takes to
complete a task using three different software programs: Program A, Program B, and Program C. They
randomly assign 30 employees to one of the programs and record the time it takes each employee to
complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or
interaction effects between the software programs and employee experience level (novice vs.
experienced). Report the F-statistics and p-values, and interpret the results.

In [11]:
#Assuming Significance level of 0.05
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Setting random seed for reproducibility
np.random.seed(123)

# Generating 2 random time samples for novice and expert
time_novice = np.random.normal(loc=15, scale=2, size=30)
time_expert = np.random.normal(loc=10, scale=2, size=30)

# Generate simulated data
data = pd.DataFrame({
    'Software': ['A']*20 + ['B']*20 + ['C']*20,
    'Experience': ['Novice']*30 + ['Experienced']*30,
    'Time': list(time_novice)+list(time_expert)
})

# Print the simulated data head 
print('Simulated Data example :')
print(data.head())

print('\n======================================================================================\n')

# Fit the two-way ANOVA model
model = ols('Time ~ C(Software) + C(Experience) + C(Software):C(Experience)', data=data).fit()
table = sm.stats.anova_lm(model, typ=1)

# Set significance level
alpha = 0.05

# Main effects and interaction effect
print(table)
print('\n')
if table['PR(>F)'][0] < alpha:
    print("Conclusion: There is a significant main effect of software.")
else:
    print("Conclusion: There is no significant main effect of software.")

if table['PR(>F)'][1] < alpha:
    print("Conclusion: There is a significant main effect of experience.")
else:
    print("Conclusion: There is no significant main effect of experience.")

if table['PR(>F)'][2] < alpha:
    print("Conclusion: There is a significant interaction effect between software and experience.")
else:
    print("Conclusion: There is no significant interaction effect between software and experience.")

Simulated Data example :
  Software Experience       Time
0        A     Novice  12.828739
1        A     Novice  16.994691
2        A     Novice  15.565957
3        A     Novice  11.987411
4        A     Novice  13.842799


                             df      sum_sq     mean_sq          F  \
C(Software)                 2.0  204.881181  102.440590  18.135666   
C(Experience)               1.0  165.079097  165.079097  29.224933   
C(Software):C(Experience)   2.0   17.481552    8.740776   1.547431   
Residual                   56.0  316.319953    5.648571        NaN   

                                 PR(>F)  
C(Software)                8.460472e-07  
C(Experience)              1.375177e-06  
C(Software):C(Experience)  2.217544e-01  
Residual                            NaN  


Conclusion: There is a significant main effect of software.
Conclusion: There is a significant main effect of experience.
Conclusion: There is no significant interaction effect between software and experience.


Here are the interpretations of the three conclusions:
1. "There is a significant main effect of software": This means that the software programs used by the employees have a significant impact on the outcome variable (e.g., completion time), independent of the experience level of the employees. This suggests that the choice of software program is an important factor that should be considered carefully when completing this task.

2. "There is a significant main effect of experience": This means that the experience level of the employees has a significant impact on the outcome variable, independent of the software program used. Specifically, this suggests that experienced employees may complete the task faster than novices, or vice versa. This finding can be helpful for the company to identify the best employees for a given task and to provide appropriate training for new employees.

3. "There is NO significant interaction effect between software and experience": This means that the effect of software on the outcome variable does not depend on the experience level of the employees, and vice versa. This suggests that the software programs perform similarly for both novices and experienced employees. This finding can be helpful for the company to decide which software program to use, as they do not need to consider the experience level of the employees when making the choice.




Q11. An educational researcher is interested in whether a new teaching method improves student test
scores. They randomly assign 100 students to either the control group (traditional teaching method) or the
experimental group (new teaching method) and administer a test at the end of the semester. Conduct a
two-sample t-test using Python to determine if there are any significant differences in test scores
between the two groups. If the results are significant, follow up with a post-hoc test to determine which
group(s) differ significantly from each other.

In [12]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind

# Setting numpy random seed
np.random.seed(45)

# Generating normal test scores with same variance for both control groups
test_score_control = np.random.normal(loc=70, scale=3, size=50)
test_score_experimental = np.random.normal(loc=85, scale=3, size=50)

# Creating the dataframe
df = pd.DataFrame({'test_score':list(test_score_control)+list(test_score_experimental),
                   'group':['control']*50 + ['experimental']*50})

# printing the sample dataframe
print('Simulated data for test_scores:')
print(df.head())
print('\n===============================\n')

null_hypothesis = "There is NO difference in test scores between the control and experimental groups."
alt_hypothesis = "There is SIGNIFICANT difference in test scores between the control and experimental groups."

# Conduct the two-sample t-test
control_scores = df[df['group'] == 'control']['test_score']
experimental_scores = df[df['group'] == 'experimental']['test_score']
t_stat, p_val = ttest_ind(control_scores, experimental_scores, equal_var=True)
print(f"t-statistic: {t_stat:.4f}, p-value: {p_val}")
print('\n')

# Significance value 
alpha = 0.05
if p_val<alpha:
    print('Reject the Null Hypothesis')
    print(f'Conclusion : {alt_hypothesis}')
else:
    print('Failed to reject the Null Hypothesis')
    print(f'Conclusion : {null_hypothesis}')

Simulated data for test_scores:
   test_score    group
0   70.079124  control
1   70.780965  control
2   68.814563  control
3   69.387097  control
4   66.185102  control


t-statistic: -28.5074, p-value: 3.096206271894725e-49


Reject the Null Hypothesis
Conclusion : There is SIGNIFICANT difference in test scores between the control and experimental groups.


In [13]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Conduct post-hoc Tukey's test
tukey_results = pairwise_tukeyhsd(df['test_score'], df['group'], 0.05)
print(tukey_results)

   Multiple Comparison of Means - Tukey HSD, FWER=0.05    
 group1    group2    meandiff p-adj  lower   upper  reject
----------------------------------------------------------
control experimental  15.8829   0.0 14.7773 16.9886   True
----------------------------------------------------------


Tukey's Results Interpretation
Reject = True suggests that there is significant difference in both control and Experimental groups also p-adj is almost 0.

Experimental group has increased the performance of test scores of students by mean of 15.88 approximately

Mean score improved by Experimental method is (14.78,16.99) with 95% confidence level



Q12. A researcher wants to know if there are any significant differences in the average daily sales of three
retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store
on those days. Conduct a repeated measures ANOVA using Python to determine if there are any significant differences in sales between the three stores. If the results are significant, follow up with a post-
hoc test to determine which store(s) differ significantly from each other.

In [14]:
import numpy as np
import pandas as pd
from statsmodels.stats.anova import AnovaRM
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# set random seed for reproducibility
np.random.seed(456)

# generate sales data for Store A, B, and C
sales_a = np.random.normal(loc=1000, scale=100, size=(30,))
sales_b = np.random.normal(loc=1050, scale=150, size=(30,))
sales_c = np.random.normal(loc=800, scale=80, size=(30,))

# create a DataFrame to store the sales data
sales_df = pd.DataFrame({'Store A': sales_a, 'Store B': sales_b, 'Store C': sales_c})

# reshape the DataFrame for repeated measures ANOVA
sales_melted = pd.melt(sales_df.reset_index(), id_vars=['index'], value_vars=['Store A', 'Store B', 'Store C'])
sales_melted.columns = ['Day', 'Store', 'Sales']

# Printing top 5 rows of generated data
print('Generated data top 5 rows : ')
print(sales_melted.head())

print('\n================================================\n')

# perform repeated measures ANOVA
rm_anova = AnovaRM(sales_melted, 'Sales', 'Day', within=['Store'])
rm_results = rm_anova.fit()
print(rm_results)

# check if null hypothesis should be rejected based on p-value
if rm_results.anova_table['Pr > F'][0] < 0.05:
    # perform post-hoc Tukey test
    print('Reject the Null Hypothesis : Atleast one of the group has different mean.\n')
    print('Tukey HSD posthoc test:')
    tukey_results = pairwise_tukeyhsd(sales_melted['Sales'], sales_melted['Store'])
    print(tukey_results)
else:
    print('NO significant difference between groups.')

Generated data top 5 rows : 
   Day    Store        Sales
0    0  Store A   933.187150
1    1  Store A   950.179048
2    2  Store A  1061.857582
3    3  Store A  1056.869225
4    4  Store A  1135.050948


               Anova
      F Value Num DF  Den DF Pr > F
-----------------------------------
Store 51.5040 2.0000 58.0000 0.0000

Reject the Null Hypothesis : Atleast one of the group has different mean.

Tukey HSD posthoc test:
    Multiple Comparison of Means - Tukey HSD, FWER=0.05    
 group1  group2  meandiff p-adj    lower     upper   reject
-----------------------------------------------------------
Store A Store B   21.2439 0.6945   -40.881   83.3688  False
Store A Store C -207.8078    0.0 -269.9328 -145.6829   True
Store B Store C -229.0517    0.0 -291.1766 -166.9268   True
-----------------------------------------------------------


Interpretation of above
In Repeated Measure ANOVA test we got p_value (Pr>F) as 0.0000 which is less than 0.05 .Reject the Null Hypothesis .Which means atleast one of the mean of groups is different.

In Tukey's Post Hoc Test we get following interpretation :

No significant difference between sales of Store A and Store B. Store B earns 21.24 dollars more than store A(becuse reject=False for this)
Significant difference between sales of Store A and Store C . Store C has approx 207.8 dollars lesser compared to store A (reject=True)
Siginficant difference between sales of Store B and Store C . Store C has approx 229.0 dollars lesser compared to store B (reject=True